Mathlib/Algebra/Algebra/Unitize.lean

```diff
@@ -0,0 +1,348 @@
+    replace hx' := congr_arg (fun y => r⁻¹ • y) hx'
+    simp only [Algebra.algebraMap_eq_smul_one, ← smul_assoc, smul_eq_mul, inv_mul_cancel hr,
+      one_smul] at hx'
+    exact hx'.symm ▸ SMulMemClass.smul_mem r⁻¹ (neg_mem a.prop)
+
+/-- If a `NonUnitalSubalgebra` over a field does not contain `1`, then its unitization is
+isomorphic to its `Algebra.adjoin`. -/
+noncomputable def NonUnitalSubalgebra.unitizationAlgEquiv {R S A : Type _} [Field R] [Ring A]
+    [Algebra R A] [SetLike S A] [NonUnitalSubringClass S A] [SMulMemClass S R A]
+    (s : S) (h1 : (1 : A) ∉ s) : Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A) :=
+  AlgEquiv.ofBijective (NonUnitalSubalgebra.unitization s)
+    ⟨NonUnitalSubalgebra.unitization_injective s h1, NonUnitalSubalgebra.unitization_surjective s⟩
+
+end Generic
+
+section Subsemiring
+
+variable {R : Type _} [NonAssocSemiring R]
+
+/-! ## Subsemirings -/
+
+
+/-- Turn a `Subsemiring` into a `NonUnitalSubsemiring` by forgetting that it contains `1`. -/
+def Subsemiring.toNonUnitalSubsemiring (S : Subsemiring R) : NonUnitalSubsemiring R :=
+  { S with carrier := S.carrier }
+
+theorem Subsemiring.one_mem_toNonUnitalSubsemiring (S : Subsemiring R) :
+    (1 : R) ∈ S.toNonUnitalSubsemiring :=
+  S.one_mem
+
+/-- Turn a non-unital subsemiring containing `1` into a subsemiring. -/
+def NonUnitalSubsemiring.toSubsemiring (S : NonUnitalSubsemiring R) (h1 : (1 : R) ∈ S) :
+    Subsemiring R :=
+  { S with
+    carrier := S.carrier
+    one_mem' := h1 }
+
+theorem Subsemiring.toNonUnitalSubsemiring_toSubsemiring (S : Subsemiring R) :
+    S.toNonUnitalSubsemiring.toSubsemiring S.one_mem = S := by cases S; rfl
+
+theorem NonUnitalSubsemiring.toSubsemiring_toNonUnitalSubsemiring (S : NonUnitalSubsemiring R)
+    (h1 : (1 : R) ∈ S) : (NonUnitalSubsemiring.toSubsemiring S h1).toNonUnitalSubsemiring = S := by
+  cases S; rfl
+
+/-- The `ℕ`-algebra equivalence between `Subsemiring.closure s` and `Algebra.adjoin ℕ s` given
+by the identity map. -/
+def Subsemiring.closureEquivAdjoinNat {R : Type _} [Semiring R] (s : Set R) :
+    Subsemiring.closure s ≃ₐ[ℕ] Algebra.adjoin ℕ s
+    where
+  toFun := Subtype.map id fun r hr => Subsemiring.closure_induction hr Algebra.subset_adjoin
+    (zero_mem _) (one_mem _) (fun _ _ => add_mem) fun _ _ => mul_mem
+  invFun := Subtype.map id fun r hr => Algebra.adjoin_induction hr Subsemiring.subset_closure
+    (natCast_mem _) (fun _ _ => add_mem) fun _ _ => mul_mem
+  -- Note: `Subtype.ext rfl` proves all of these, but, depressingly, it's too slow.
+  left_inv _ := Subtype.ext <| by simp only [Subtype.map_coe, id_eq]
+  right_inv _ := Subtype.ext <| by simp only [Subtype.map_coe, id_eq]
+  map_mul' _ _ := Subtype.ext <| by
+    simp only [Subtype.map_coe, Submonoid.coe_mul, coe_toSubmonoid, id_eq,
+      Subalgebra.coe_toSubsemiring]
+  map_add' _ _ := Subtype.ext <| by
+    simp only [Subtype.map_coe, coe_add, id_eq, Subalgebra.coe_toSubsemiring]
+  commutes' _ := Subtype.ext rfl
```

Comment:
Aha, it already exists! But there's a name mismatch: [Subalgebra.equivOfEq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#Subalgebra.equivOfEq)

Reply:
But we can't use it anyway because `Subsemiring.closure s : Subsemiring R`, it does not have type `Subalgebra ℕ R`. We could write it so that it takes `SetLike` stuff, but even that wouldn't work because `SMulMemClass` has the scalars as an `outParam`. I experimented a bit ago with making that a `semiOutParam` instead, which would allow this kind of thing to work, and the breakage wasn't too severe. But I think in this case the extra `Subtype.ext` boilerplate is not too cumbersome.

Reply:
It turns out (see suggestions below) that it works perfectly fine using [subalgebraOfSubsemiring](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#subalgebraOfSubsemiring) and [subalgebraOfSubring](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#subalgebraOfSubring)